go/enclave/components/rollup_compression.go

```diff
@@ -0,0 +1,506 @@
+	return &RollupCompression{
+		batchRegistry:          batchRegistry,
+		batchExecutor:          batchExecutor,
+		dataEncryptionService:  dataEncryptionService,
+		dataCompressionService: dataCompressionService,
+		storage:                storage,
+		chainConfig:            chainConfig,
+		logger:                 logger,
+	}
+}
+
+// temporary data structure to help build a batch from the information found in the rollup
+type batchFromRollup struct {
+	transactions []*common.L2Tx
+	seqNo        *big.Int
+	height       *big.Int
+	txHash       gethcommon.Hash
+	time         uint64
+	l1Proof      common.L1BlockHash
+
+	header *common.BatchHeader // for reorgs
+}
+
+// CreateExtRollup - creates a compressed and encrypted External rollup from the internal data structure
+func (rc *RollupCompression) CreateExtRollup(r *core.Rollup) (*common.ExtRollup, error) {
+	header, err := rc.createRollupHeader(r.Batches)
+	if err != nil {
+		return nil, err
+	}
+	encryptedHeader, err := rc.serialiseCompressAndEncrypt(header)
+	if err != nil {
+		return nil, err
+	}
+
+	transactions := make([][]*common.L2Tx, len(r.Batches))
+	for i, batch := range r.Batches {
+		transactions[i] = batch.Transactions
+	}
+	encryptedTransactions, err := rc.serialiseCompressAndEncrypt(transactions)
+	if err != nil {
+		return nil, err
+	}
+
+	return &common.ExtRollup{
+		Header:               r.Header,
+		BatchPayloads:        encryptedTransactions,
+		CalldataRollupHeader: encryptedHeader,
+	}, nil
+}
+
+// ProcessExtRollup - given an External rollup, responsible with checking and saving all batches found inside
+func (rc *RollupCompression) ProcessExtRollup(rollup *common.ExtRollup) error {
+	transactionsPerBatch := make([][]*common.L2Tx, 0)
+	err := rc.decryptDecompressAndDeserialise(rollup.BatchPayloads, &transactionsPerBatch)
+	if err != nil {
+		return err
+	}
+
+	calldataRollupHeader := new(common.CalldataRollupHeader)
+	err = rc.decryptDecompressAndDeserialise(rollup.CalldataRollupHeader, calldataRollupHeader)
+	if err != nil {
+		return err
+	}
+
+	// The recreation of batches is a 2-step process:
+
+	// 1. calculate fields like: sequence, height, time, l1Proof, from the implicit and explicit information from the metadata
+	incompleteBatches, err := rc.createIncompleteBatches(calldataRollupHeader, transactionsPerBatch, rollup.Header.L1Proof)
+	if err != nil {
+		return err
+	}
+
+	// 2. execute each batch to be able to calculate the hash which is necessary for the next batch as it is the parent.
+	err = rc.executeAndSaveIncompleteBatches(calldataRollupHeader, incompleteBatches)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// the main logic that goes from a list of batches to the rollup header
+func (rc *RollupCompression) createRollupHeader(batches []*core.Batch) (*common.CalldataRollupHeader, error) {
+	reorgs := make([]*common.BatchHeader, len(batches))
+
+	deltaTimes := make([]*big.Int, len(batches))
+	startTime := batches[0].Header.Time
```

Comment:
Might be worth checking that the batches list isn't empty here? Avoid panic risk/give back a useful error if things get weird.

Reply:
there is such a check when the "Rollup" is created.